lager/lenses.hpp

```diff
@@ -123,51 +129,169 @@ auto attr(Member member)
+        return [&, f = LAGER_FWD(f)](auto&& whole) {
+            using T = std::decay_t<decltype(whole.value())>;
+            return f(LAGER_FWD(whole).value_or(T{}))(
+                [&](auto&& x) { return LAGER_FWD(x); });
+        };
+    });
+}
+
+namespace detail {
+
+template <
+    template <typename>
+    typename PartMeta,
+    typename Whole,
+    typename Lens,
+    typename Functor>
+decltype(auto) opt_impl(Whole&& whole, Lens&& lens, Functor&& f) {
+    using Part = typename PartMeta<std::decay_t<decltype(::lager::view(
+        std::forward<Lens>(lens),
+        std::declval<std::decay_t<decltype(whole.value())>>()))>>::type;
+
+    if (whole.has_value()) {
+        return std::invoke(
+            std::forward<Functor>(f),
+            Part{::lager::view(
+                std::forward<Lens>(lens), std::forward<Whole>(whole).value())})(
+            [&](Part part) {
+                if (part.has_value()) {
+                    return std::decay_t<Whole>{::lager::set(
+                        std::forward<Lens>(lens),
+                        std::forward<Whole>(whole).value(),
+                        std::move(part).value())};
                 } else {
-                    return std::forward<decltype(p)>(p);
+                    return std::forward<Whole>(whole);
                 }
             });
+    } else {
+        return std::invoke(std::forward<Functor>(f), Part{std::nullopt})(
+            [&](auto&&) { return std::forward<Whole>(whole); });
+    }
+}
+
+template <typename T> struct remove_opt { using type = T; };
+template <typename T> struct remove_opt<std::optional<T>> { using type = T; };
+template <typename T> using remove_opt_t = typename remove_opt<T>::type;
+
+template <typename T> struct to_opt {
+    using type = std::optional<remove_opt_t<std::decay_t<T>>>;
+};
+
+template <typename T> struct add_opt {
+    using type = std::optional<std::decay_t<T>>;
+};
+
+} // namespace detail
+
+/*!
+ * `Lens<W, P> -> Lens<[W], [P]>`
+ */
+template <typename Lens>
+auto optmap(Lens&& lens) {
+    return zug::comp([lens = std::forward<Lens>(lens)](auto&& f) {
+        return [&, f = LAGER_FWD(f)](auto&& whole) {
+            return detail::opt_impl<detail::add_opt>(LAGER_FWD(whole), lens, f);
+        };
+    });
+}
+
+/*!
+ * `Lens<W, [P]> -> Lens<[W], [P]>`
+ */
+template <typename Lens>
+auto optbind(Lens&& lens) {
+    return zug::comp([lens = std::forward<Lens>(lens)](auto&& f) {
+        return [&, f = LAGER_FWD(f)](auto&& whole) {
+            return detail::opt_impl<zug::meta::identity>(LAGER_FWD(whole), lens, f);
         };
     });
 }
 
+/*!
+ * `(Lens<W, P> | Lens<W, [P]>) -> Lens<[W], [P]>`
+ */
+template <typename Lens>
+auto optlift(Lens&& lens) {
+    return zug::comp([lens = std::forward<Lens>(lens)](auto&& f) {
+        return [&, f = LAGER_FWD(f)](auto&& whole) {
+            return detail::opt_impl<detail::to_opt>(LAGER_FWD(whole), lens, f);
+        };
+    });
+}
+
+
 } // namespace lens
 
```

Comment:
This file is growing... Maybe we should split it into multiple files in `lager/lens/...`.

Reply:
agreed.